Computation of secondary moments, noise correction of RhoHV and pseudo-PPI from RHI volume

jfigui opened the pull request:
Dear Py-ART users,

I have added the following functionalities to Py-ART:
- A correction of RhoHV for noise in correct/noise.py
- The computation of the new fields: noise in dBZ, signal to noise ratio, logarithmic RhoHv (L parameter), and circular depolarization ratio in retrieve/simple_moment_calculations.
- A cross-section of a volume of RHIs along particular azimuths (pseudo-PPI) in util/xsect.py

@jjhelmus, I think we should set standard names for the new fields.

Let me know if you think these new functionalities can be implemented in Py-ART and are up to standard.

Greetings from Switzerland,


Jordi

josephhardinee commented on pyart/correct/noise.py:
```diff
@@ -0,0 +1,99 @@
+    fill_value = urhohv.get_fill_value()
+
+    rhohv_data = np.ma.masked_where(
+        mask, urhohv*np.ma.sqrt((1.+1./snr_h)*(1.+zdr/(alpha*snr_h))))
```
Can you provide a reference in the literature to the source of this equation?

jfigui replied:
Hi @josephhardinee!

It is all about playing with signal and noise to subtract. I am not sure we ever published this one formally but I will try to see in my archives if I can find something.

Thanks for the comment,


Jordi

jjhelmus commented on pyart/correct/noise.py:
```diff
@@ -0,0 +1,99 @@
+    """
+    # parse the field parameters
+    if urhohv_field is None:
+        rhohv_field = get_field_name('uncorrected_cross_correlation_ratio')
```
Should this be urhohv_field = get_field_name('uncorrected_cross_correlation_ratio')

jjhelmus commented on pyart/correct/noise.py:
```diff
@@ -0,0 +1,99 @@
+    rhohv_data = np.ma.masked_where(
+        mask, urhohv*np.ma.sqrt((1.+1./snr_h)*(1.+zdr/(alpha*snr_h))))
+    rhohv_data.set_fill_value(fill_value)
+    rhohv_data.data[mask.nonzero()] = fill_value
```
This destroys the masking of `rhohv_data` setting them to fill_value, is there a reason why the data should be un-masked?  Can the following be used instead?  

rhohv_data = urhohv*np.ma.sqrt((1.+1./snr_h)*(1.+zdr/(alpha*snr_h)))

jjhelmus commented on pyart/correct/noise.py:
```diff
@@ -0,0 +1,99 @@
+    rhohv_data.set_fill_value(fill_value)
+    rhohv_data.data[mask.nonzero()] = fill_value
+    is_above1 = rhohv_data > 1.
+    rhohv_data[is_above1.nonzero()] = 1.
```
The nonzero() call is not needed.  In fact `rho_hv[rhohv_data > 1] = 1` will work fine

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+
+    Returns
+    -------
+    noisedBZ : dictionary of dictionaries
```
This is just a dictionary ` noisedBZ : dict`.

jjhelmus commented on pyart/correct/noise.py:
```diff
@@ -0,0 +1,99 @@
+
+    Returns
+    -------
+    rhohv : dictionary of dictionaries
```
This is a dictionary not a dictionary of dictionaries.  `rhohv : dict`

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+
+    Returns
+    -------
+    snr : dictionary of dictionaries
```
dictionary

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+    mask = np.ma.getmaskarray(refl)
+    fill_value = refl.get_fill_value()
+
+    snr_data = np.ma.masked_where(mask, refl-noisedBZ)
```
refl-noiseDBZ will take care of the masking.

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+    is_one = rhohv >= 1.
+    rhohv[is_one.nonzero()] = 0.9999
+
+    l_data = np.ma.masked_where(mask, -np.ma.log10(1.-rhohv))
```
l_data = -np.ma.log10(1.-rhohv) is sufficient.

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+    mask = np.ma.getmaskarray(rhohv)
+    fill_value = rhohv.get_fill_value()
+
+    cdr_data = np.ma.masked_where(
```
The use of np.ma.masked_where is not needed.

jjhelmus commented on pyart/util/xsect.py:
```diff
@@ -101,6 +102,89 @@ def cross_section_ppi(radar, target_azimuths):
+            ray_number = np.argmin(np.abs(sweep_elevations - target_elevation))
+            pppi_rays.append(ray_number + sweep_slice.start)
+
+    _range = _copy_dic(radar.range)
```
The lines from here on seems to be nearly identical to those in `cross_section_ppi`.  Would it make sense to make these into a private function which both call?

jjhelmus commented:
@jfigui Thanks for the new features!  I made a few comment on suggested changes.  Mostly they deal with what looks to be a bit of a misunderstanding on how [masked arrays](http://docs.scipy.org/doc/numpy/reference/maskedarray.generic.html) work.  Typically masked arrays can be treated just like normal ndarray objects except that `np.ma` functions should be used (np.ma.power, np.ma.sin, etc).  The masking of the results is automatically taken care of by NumPy.

I would really appreciate some unit tests for these new function.

I agree that some new fields need to be added to the default config, feel free to add these.

The CI is failing for a unrelated reason which has been fixed in #586.

jfigui commented:
@jjhelmus thanks for checking the code. Indeed looks like my understanding of masked arrays is still a bit fuzzy.
I will correct the functions according to your suggestions and try to create some tests functions in the coming days.
I do think that repeated code in cross_section_xxx should be placed in a private function. I will do that.

jjhelmus commented:
Great @jfigui, let me know if you need any help with this.

jfigui commented:
@jjhelmus it just crossed my mind that the xsect functions look for the nearest neighbour without bothering to check if it is close enough. I would recommend to allow for a user defined angle distance tolerance (if angle distance is None the function would behave as it is now). What do you think about the idea?

jjhelmus commented:
I like the idea.  I would probably make sense to issue a warning when a angle is dropped because it fails the distance tolerance.  Feel free to either add that to this PR or submit a second PR with the change.

jfigui commented:
@jjhelmus, I have implemented the changes in the code but I am busy doing other stuff and I did not have the time to prepare the testing functions. Probably next week I will have some time to spend on this issue but I think I will need your help for the ingestion of test data.
Do you prefer that I first submit the changes in the code or I wait until I have also the testing functions ready?

jjhelmus commented:
@jfigui Go ahead and submit the changes, tests can be added later.  I might even be able to find some time in the next week to write them myself.

jfigui commented:
@jjhelmus, I finally managed to submit the requested corrections. In addition to that:
I renamed correct/noise.py to correct/bias_and_noise.py and added a function that corrects the bias (meant for reflectivity and differential reflectivity bias correction)
I found out there was a bug in join_radar at utils/radar_utils.py. It was not picking up the right fill_value so I corrected that.
I corrected rainbow_wrl.py to get the azimuth between [0, 360] and make the frequency an array as discussed in another thread.
Let me know if now the things are up to standard.

Greetings from Switzerland! (which is getting busier in the Open Source :))

jjhelmus commented on pyart/correct/bias_and_noise.py:
```diff
@@ -0,0 +1,155 @@
+"""
+pyart.correct.noise
```
Should be pyart.correct.bias_and_noise

jjhelmus commented on pyart/correct/bias_and_noise.py:
```diff
@@ -0,0 +1,155 @@
+    alpha = np.ma.power(10., 0.1*(nh-nv))
+
+    mask = np.ma.getmaskarray(urhohv)
+    fill_value = urhohv.get_fill_value()
```
This can fail if urhohv is an ndarray not a masked array.  Can the `get_fillvalue` function be used here?

jjhelmus commented on pyart/correct/bias_and_noise.py:
```diff
@@ -0,0 +1,155 @@
+    rhohv_data = urhohv*np.ma.sqrt((1.+1./snr_h)*(1.+zdr/(alpha*snr_h)))
+    rhohv_data[rhohv_data > 1.] = 1.
+    rhohv_data.set_fill_value(fill_value)
+    rhohv_data.data[mask.nonzero()] = fill_value
```
This should be `rhohv_data.data[mask] = np.ma.masked`  Additional I do not think setting the fill_value is really needed.

jjhelmus replied:
Sorry missed that this is using `.data`.  Setting .data should not be done as it is not preserved.  Use `rhohv_data[mask] = np.ma.masked` to mask values.  Modifying `.data` and expecting the mask to be updates is not a good idea.

jjhelmus commented on pyart/correct/bias_and_noise.py:
```diff
@@ -0,0 +1,155 @@
+        raise KeyError('Field not available: ' + field_name)
+
+    mask = np.ma.getmaskarray(field_data)
+    fill_value = field_data.get_fill_value()
```
Same as above, I do not think this is need.

jjhelmus commented on pyart/correct/bias_and_noise.py:
```diff
@@ -0,0 +1,155 @@
+
+    corr_field_data = field_data - bias
+    corr_field_data.set_fill_value(fill_value)
+    corr_field_data.data[mask] = fill_value
```
`corr_field_data.data[mask] = np.ma.masked` and remove the line which sets the fill_value.

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+
+    noisedBZ = get_metadata(noise_field)
+    noisedBZ['data'] = np.tile(noisedBZ_vec, (nrays, 1))
+    noisedBZ['data'].set_fill_value(get_fillvalue())
```
This is fine but probably is not needed.

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+        raise KeyError('Field not available: ' + noise_field)
+
+    mask = np.ma.getmaskarray(refl)
+    fill_value = refl.get_fill_value()
```
`refl` may not have a `get_fill_value()` method.  Use the `get_fillvalue()` method or remove.

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+
+    snr_data = refl-noisedBZ
+    snr_data.set_fill_value(fill_value)
+    snr_data.data[mask.nonzero()] = fill_value
```
`snr_data[mask] = np.ma.masked` although this should have propagated from the refl array automatically.

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+        raise KeyError('Field not available: ' + rhohv_field)
+
+    mask = np.ma.getmaskarray(rhohv)
+    fill_value = rhohv.get_fill_value()
```
Remove or use `get_fillvalue()`

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+    rhohv[rhohv >= 1.] = 0.9999
+    l_data = -np.ma.log10(1.-rhohv)
+    l_data.set_fill_value(fill_value)
+    l_data.data[mask.nonzero()] = fill_value
```
`l_data[mask] = np.ma.masked`

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+    zdr = np.ma.power(10., 0.1*zdrdB)
+
+    mask = np.ma.getmaskarray(rhohv)
+    fill_value = rhohv.get_fill_value()
```
Same as above.

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,200 @@ def calculate_snr_from_reflectivity(
+            (1.+1./zdr-2.*rhohv*np.ma.sqrt(1./zdr)) /
+            (1.+1./zdr+2.*rhohv*np.ma.sqrt(1./zdr))))
+    cdr_data.set_fill_value(fill_value)
+    cdr_data.data[mask.nonzero()] = fill_value
```
`cdr_data[mask] = np.ma.masked`

jjhelmus commented on pyart/util/xsect.py:
```diff
@@ -41,64 +46,176 @@ def cross_section_ppi(radar, target_azimuths):
+            else:
+                d_az_min = np.min(d_az)
+                if d_az_min > az_tol:
+                    print('WARNING: No azimuth found whithin tolerance ' +
```
This should use Python's built in `warnings` module so that this can be logged or caught by the user.  `warnings.warn` is the function to use.

jjhelmus commented on pyart/util/xsect.py:
```diff
@@ -41,64 +46,176 @@ def cross_section_ppi(radar, target_azimuths):
+
+    rhi_nsweeps = len(valid_azimuths)
+    if rhi_nsweeps == 0:
+        print('WARNING: No azimuth found whithin tolerance.')
```
Perhaps this should raise an Error instead of returning None.

jjhelmus commented on pyart/util/xsect.py:
```diff
@@ -41,64 +46,176 @@ def cross_section_ppi(radar, target_azimuths):
+            else:
+                d_el_min = np.min(d_el)
+                if d_el_min > el_tol:
+                    print('WARNING: No elevation found whithin tolerance ' +
```
same as above

jjhelmus commented on pyart/util/xsect.py:
```diff
@@ -41,64 +46,176 @@ def cross_section_ppi(radar, target_azimuths):
+
+    ppi_nsweeps = len(valid_elevations)
+    if ppi_nsweeps == 0:
+        print('WARNING: No elevation found whithin tolerance.')
```
same as above

jjhelmus commented:
@jfigui Added a few new comments.  Most of them are about setting `fill_value` and the `data` attribute of masked arrays.  

In general one should not depend on the `fill_value` of a masked array being correct as this value is not guaranteed to propagate.  

Additionally setting the `data` value for elements of a masked array which are masked in pointless.  NumPy assumes that these values can take on arbitrary values since the values are masked and does not attempt to preserve these values when operations are performed on the masked array.

Also should mention that it best to separate the addition of new functionality into different PR.  Each PR should only add one new feature or fix one particular bug.  This makes reviewing and merging easier.  Happy to get this one through as-is just something to keep in mind for the future.

meteoswiss-mdr commented:
@jjhelmus,
I checked the issue and the fill_value is not propagated (it takes the default 1e+20 for float arrays). So I think we should explicitly set the fill_value. There may be contexts where it can get handy to know the fill_value directly from the data. What do you think?
The rest of your comments I have taken them on board. I agree with you that as a general rule each functionality should have a different PR but it was such a minor addition that I though there was no need to bother.
I will commit the changes in a few minutes.

Greetings from Switzerland,

Jordi

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,190 @@ def calculate_snr_from_reflectivity(
+
+    noisedBZ = get_metadata(noise_field)
+    noisedBZ['data'] = np.tile(noisedBZ_vec, (nrays, 1))
+#    noisedBZ['data'].set_fill_value(get_fillvalue())
```
Can you remove this line, dead code is not good to include.

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,190 @@ def calculate_snr_from_reflectivity(
+        raise KeyError('Field not available: ' + noise_field)
+
+    snr_data = refl-noisedBZ
+#    snr_data.set_fill_value(get_fillvalue())
```
Remove

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,190 @@ def calculate_snr_from_reflectivity(
+    else:
+        raise KeyError('Field not available: ' + rhohv_field)
+
+#    mask = np.ma.getmaskarray(rhohv)
```
Remove

jjhelmus commented on pyart/retrieve/simple_moment_calculations.py:
```diff
@@ -66,3 +70,190 @@ def calculate_snr_from_reflectivity(
+    rhohv[rhohv >= 1.] = 0.9999
+    l_data = -np.ma.log10(1.-rhohv)
+#    l_data.data[mask] = np.ma.masked
+#    l_data.set_fill_value(get_fillvalue())
```
Remove

jjhelmus commented:
@jfigui I'm going to remove the commented out lines of code and merge this, thanks for putting this together and seeing it through.  The CI failures are unrelated to this change and I'll investigate shortly.

jjhelmus commented:
> I checked the issue and the fill_value is not propagated (it takes the default 1e+20 for float arrays). So I think we should explicitly set the fill_value. There may be contexts where it can get handy to know the fill_value directly from the data. What do you think?

The issue I was referring to is that the fill_value is not propogated when masked arrays are operated on.  For example, see how the fill_value of the `output` variable varies below.

```
import numpy as np
a = np.ma.array([1,2,3], mask=[False, False, True], fill_value=999)
b = np.ma.array([1,2,3], mask=[False, False, True], fill_value=888)
c = np.ma.array([1,2,3])

output = a + b  # output.fill_value is 999
output = b + a  # output.fill_value is 888
output = a + c  # output.fill_value is 999
output = c + a  # output.fill_value is 999999 (sometimes...)
```

I agree that setting the `fill_value` of a masked array is prudent, but I would not rely on it being correctly set when retrieving the value.  Using `np.ma.filled` with an explicit `fill_value` parameter is a more reliable method for creating a ndarray with a particular value where the original data is masked.

meteoswiss-mdr commented:
@jjhelmus, thank you for your patience.

Now that I have a better understanding of masked arrays I agree with you. There is no use on setting the fill_value and the data value of masked entries and whenever needed filled should be used.

Have a nice day,


Jordi
